packages/datastore/src/datastore.ts

```diff
@@ -287,14 +292,21 @@ class Datastore implements IIterable<Table<Schema>>, IMessageHandler, IDisposabl
    *
    * @returns A promise which resolves when the action is complete.
    *
-   * @throws An exception if `undo` is called during a mutation.
+   * @throws An exception if `undo` is called during a mutation, or if no
+   *   server adapter has been set for the datastore.
    *
    * #### Notes
    * If changes are made, the `changed` signal will be emitted before
    * the promise resolves.
    */
   undo(transactionId: string): Promise<void> {
-    throw Error('Undo is not implemented');
+    if (!this._adapter) {
+      throw Error('No server adapter has been set for the datastore');
+    }
+    if (this.inTransaction) {
+      throw Error('Cannot undo during a transaction');
```

Comment:
Why not though? Is this purely conceptual (this *shouldn't* happen), or technical (this *mustn't* happen)?

Reply:
While it might technically work, I think it's worth enforcing some manner of mental consistency. "Just because we can, doesn't mean we should".

Reply:
Yeah, I was mostly trying to make the note in the docstring true here, and it seemed like a reasonable restriction.

Reply:
Sure, this was mostly to ensure we were all on the same page. Consider it resolved then 👍